crates/red_knot_python_semantic/src/visibility_constraints.rs

```diff
@@ -184,49 +183,121 @@ const MAX_RECURSION_DEPTH: usize = 24;
 /// for a particular [`Constraint`], if your formula needs to consider how a particular runtime
 /// property might be different at different points in the execution of the program.
 #[derive(Clone, Debug, PartialEq, Eq)]
-pub(crate) enum VisibilityConstraint<'db> {
+pub(crate) struct VisibilityConstraint<'db>(VisibilityConstraintInner<'db>);
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(crate) enum VisibilityConstraintInner<'db> {
```

Comment:
Adding this as a separate type made it easier to verify that no callers are constructing this manually, and always use one of the builder's constructor methods